conda.rb

```diff
@@ -1,93 +1,83 @@
+# coding: utf-8
 # frozen_string_literal: true
 
 require "httparty"
 require "json"
-require "msgpack"
 require "singleton"
 
 class Conda
   include Singleton
 
-  CHANNELS = [
-    # channel, domain, directory_path_to channeldata.json
-    ["pkgs/main", "repo.anaconda.com"],
-    # TODO: enable me when asked to.
-    # ["conda-forge", "conda.anaconda.org", "/conda-forge"],
-  ].freeze
+  attr_reader :main
 
   def initialize
-    @redis = if ENV["RACK_ENV"] == "test"
-               MockRedis.new
-             else
-               Redis.new(url: ENV["REDIS_SERVER"], driver: :hiredis)
-             end
+    @main = Conda::Channel.new("pkgs/main", "repo.anaconda.com")
   end
 
-  ###########
-  # Getters # -> Used for Api Calls
-  ###########
+  class Channel
+    ARCHES = %w[linux-64 osx-64 win-64 noarch].freeze
 
-  def package_names
-    @redis.smembers("package_names").sort
-  end
-
-  def package(channel, name)
-    pack = @redis.get("packages:#{channel}/#{name}")
-    return unless pack
+    attr_reader :timestamp, :packages
 
-    MessagePack.unpack(pack.force_encoding("ASCII-8BIT")).update({ "name" => name })
-  end
+    def initialize(channel, domain)
+      @channel = channel
+      @domain = domain
+      @timestamp = Time.now
+      @mutex = Mutex.new
+      reload
+    end
 
-  def latest(count)
-    packages = @redis.scan_each(match: "packages:*").to_a.map do |key|
-      channel, name = channel_and_name_from_key(key)
-      package = package(channel, name)
-      next if package["timestamp"].nil?
+    def reload
+      @mutex.synchronize { @packages = retrieve_packages }
```

Comment:
Won't we also need to protect reading of this while we're writing? It looks like rufus winds up bringing in concurrent-ruby 4 levels into the dep tree, let's maybe add it at the top level and use a `Concurrent::ReadWriteLock`?

[Chatted about this f2f]